filament/src/components/TransformManager.cpp

```diff
@@ -422,6 +440,31 @@ void FTransformManager::computeWorldTransform(
     }
 }
 
+void FTransformManager::computeWorldOrientation(
+        math::mat3f& outOrientation, 
+        math::mat3f const& parentOrientation, 
+        math::mat3f const& localOrientation) {
+
+    outOrientation = parentOrientation * localOrientation;
+}
+
+math::mat3f FTransformManager::getCompoundOrientation(Instance ci) const noexcept {
+    const mat4f& world = mManager[ci].world;
+    const mat3f& orientation = mManager[ci].orientation;
+
+    // we need to strip everything from our world transform, except rotation
+    mat3f worldRotation = mat3f(world[0].xyz, world[1].xyz, world[2].xyz);
+    worldRotation[0] = normalize(worldRotation[0]);
+    worldRotation[1] = normalize(worldRotation[1]);
+    worldRotation[2] = normalize(worldRotation[2]);
+    // we need to invert the rotation, because we undo it to keep texture direction
+    // orthonormal matrix -> transpose = inverse
+    worldRotation = transpose(worldRotation);
+
+    // we apply the orientation transformation first, then our world rotation
+    return worldRotation * orientation;
```

Comment:
So we need this "tweaked" matrix (inverse world rotation and forward material orientation combined) because we can only access fragment position in world-space?

Reply:
Exactly. Our input position is already transformed into world space, and material orientation is computed in local/object space. Thus we need to undo the world transformation's rotation, in order to locally apply the orientation.

Reply:
I realized multiplication should happen in the reverse order. Fixed in eb6e994